slurm/train.slurm

```diff
@@ -61,6 +56,83 @@ echo "Gradient accumulation steps: $GRAD_ACC_STEPS"
 MASTER_ADDR=$(scontrol show hostnames $SLURM_JOB_NODELIST | head -n 1)
 MASTER_PORT=6000
 
+# Get all hostnames in allocation
+HOSTNAMES=($(scontrol show hostnames $SLURM_JOB_NODELIST))
+NUM_HOSTS=${#HOSTNAMES[@]}
+LAST_NODE=${HOSTNAMES[-1]}
+LAST_NODE_IP=$(getent hosts $LAST_NODE | awk '{ print $1 }')
+
+# Default: all nodes are training nodes
+TRAIN_NODES=("${HOSTNAMES[@]}")
+
+if [[ "$USE_VLLM" == "true" ]]; then
+    WORLD_SIZE=$((WORLD_SIZE - GPUS_PER_NODE))
+    MODEL=$(grep 'model_name_or_path:' $CONFIG_FILE | awk '{print $2}')
+    REVISION=$(grep 'model_revision:' $CONFIG_FILE | awk '{print $2}')
+
+    # Some models like Qwen have a number of attention heads that is not divisible by the default tensor parallel size, so we need to adjust that
+    TP=$(python3 - <<EOF
```

Comment:
This is ugly as hell, so could be optionally ported to a Python script we run to get the TP value